src/onnx/parse_scan.cpp

```diff
@@ -0,0 +1,314 @@
+                           "} would exceed the maximum iteration limit {" +
+                           std::to_string(parser.max_loop_iterations) + "}");
+
+        // Check that state variable shapes match between the Scan node and its body attribute
+        for(auto i = 0; i < n; ++i)
+            if(args[i]->get_shape() != body_params[i]->get_shape())
+                MIGRAPHX_THROW("Scan: State input " + std::to_string(i) + " shape {" +
+                               to_string(args[i]->get_shape()) +
+                               "} does not match corresponding body input shape {" +
+                               to_string(body_params[i]->get_shape()) + "}");
+
+        // Check that the shapes of scan inputs sliced across scan input axes match the shapes of
+        // the body attribute scan inputs
+        for(auto i = 0; i < m; ++i)
+        {
+            auto node_shape = args[i + n]->get_shape();
+            auto node_lens  = node_shape.lens();
+            node_lens.erase(node_lens.begin() + scan_input_axes[i]);
+            auto slice_sh = shape(node_shape.type(), std::move(node_lens));
+            if(body_params[i + n]->get_shape() != slice_sh)
+                MIGRAPHX_THROW("Slice: Sliced scan input " + std::to_string(i) + " shape {" +
+                               to_string(slice_sh) +
+                               "} does not match corresponding body input shape {" +
+                               to_string(body_params[i + n]->get_shape()) + "}");
+        }
+
+        modify_body(body, args, n, m, scan_input_axes, scan_input_directions);
+
+        auto max_iter_lit = info.add_literal(literal{shape{shape::int64_type}, {num_iters}});
+        auto cond_lit     = info.add_literal(literal{shape{shape::bool_type}, {true}});
+        std::vector<instruction_ref> loop_args{max_iter_lit, cond_lit};
+        loop_args.insert(loop_args.end(), args.begin(), args.begin() + n);
+
+        auto loop =
+            info.add_instruction(make_op("loop",
+                                         {{"max_iterations", num_iters},
+                                          {"scan_output_directions", scan_output_directions}}),
+                                 loop_args,
+                                 {body});
+
+        std::vector<instruction_ref> ret;
+        ret.reserve(n + k);
+        for(auto i = 0; i < n; ++i)
+            ret.push_back(info.add_instruction(make_op("get_tuple_elem", {{"index", i}}), loop));
+
+        for(auto i = 0; i < k; ++i)
+        {
+            auto o = info.add_instruction(make_op("get_tuple_elem", {{"index", i + n}}), loop);
+            // Loop scan_outputs are concatenated along axis 0, so it must be transposed to the
+            // index specified by the corresponding scan_output_axis
+            auto perm = make_perm_for_scan_out(o->get_shape().ndim(), scan_output_axes[i]);
+            ret.push_back(info.add_instruction(make_op("transpose", {{"permutation", perm}}), o));
```

Comment:
I am missing something here. Can you explain why the transpose is necessary ? Transpose will probably not result in correct element order.

Reply:
I've expanded the comment, hopefully it's more helpful now.
There are a couple of test cases that cover this scenario, and the correct result is produced.